<a href="https://colab.research.google.com/github/Harsha2795-tech/my-github-assignment/blob/main/HarshaNandala_Liveproject1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np

In [2]:
from google.colab import files
uploaded = files.upload()

Saving bangalore_tech_salaries.csv to bangalore_tech_salaries.csv


In [3]:
df = pd.read_csv("bangalore_tech_salaries.csv")

In [ ]:
df.head()

In [ ]:
df.shape

In [ ]:
df.columns

In [ ]:
df.dtypes

In [ ]:
df.isnull().sum()

In [ ]:
df.duplicated().sum()

In [ ]:
df.describe()

# This cell loads the dataset and inspects its structure, missing values,
# data types, and duplicates before data cleaning.

In [4]:
# Task 1: Load and Inspect the Dataset
print("========== DATASET SUMMARY ==========")
print("Rows:", df.shape[0])
print("Columns:", df.shape[1])
print("Duplicate Rows:", df.duplicated().sum())

print("\nMissing Values:")
print(df.isnull().sum())

print("\nData Types:")
print(df.dtypes)

========== DATASET SUMMARY ==========
Rows: 1015
Columns: 12
Duplicate Rows: 15

Missing Values:
Employee_ID         0
Role                0
years_exp           0
Current_CTC         0
Previous_CTC      200
Company             0
company_TYPE        0
Skills             27
Location           20
Education_Tier      0
Joining_Year        0
Work_Mode           0
dtype: int64

Data Types:
Employee_ID       object
Role              object
years_exp          int64
Current_CTC       object
Previous_CTC      object
Company           object
company_TYPE      object
Skills            object
Location          object
Education_Tier    object
Joining_Year       int64
Work_Mode         object
dtype: object


# Task 2: Data Cleaning

# Clean the dataset by standardizing column names, role names,
# CTC values, company type, education tier, work mode,
# removing duplicates, and handling missing values.

In [8]:
df.columns = (
    df.columns
      .str.strip()
      .str.lower()
      .str.replace(" ", "_")
)

In [9]:
sorted(df["role"].unique())

['Analytics Engineer',
 'BA',
 'BE',
 'BI Analyst',
 'Backend Dev',
 'Backend Developer',
 'Backend Engineer',
 'Business Analyst',
 'Business Systems Analyst',
 'DA',
 'DS',
 'Data Analyst',
 'Data Science Engineer',
 'Data Scientist',
 'Designer',
 'DevOps',
 'DevOps Engineer',
 'FE',
 'Frontend Dev',
 'Frontend Developer',
 'Frontend Engineer',
 'Full Stack Engineer',
 'FullStack',
 'Fullstack Dev',
 'Infra Engineer',
 'ML Engineer',
 'PM',
 'Product Designer',
 'Product Lead',
 'Product Manager',
 'SDE Backend',
 'SDE FS',
 'SDE Frontend',
 'SDE Full-Stack',
 'SDE-Backend',
 'SDE-Frontend',
 'SRE',
 'Site Reliability Engineer',
 'Sr PM',
 'UI Designer',
 'UI/UX',
 'UX Designer']

In [10]:
role_mapping = {
    "Backend Engineer": "SDE Backend",
    "Backend Dev": "SDE Backend",
    "BE": "SDE Backend",
    "SDE Backend": "SDE Backend",

    "Frontend Engineer": "SDE Frontend",
    "Frontend Dev": "SDE Frontend",
    "SDE Frontend": "SDE Frontend",

    "Full Stack Engineer": "SDE Full-Stack",
    "Full Stack Developer": "SDE Full-Stack",
    "SDE Full Stack": "SDE Full-Stack"
}

In [11]:
df["role_clean"] = df["role"].replace(role_mapping)

In [12]:
df["role_clean"].value_counts()

,count
role_clean,
SDE Backend,79
SDE Frontend,57
BA,43
Data Scientist,40
SDE Full-Stack,38
Business Analyst,37
Site Reliability Engineer,33
ML Engineer,32
Business Systems Analyst,32


In [13]:
def parse_ctc(value):

    if pd.isna(value):
        return np.nan

    value = str(value).replace("₹", "").replace(",", "").replace("LPA", "").strip()

    try:
        number = float(value)

        if number > 100000:
            number = number / 100000

        return number

    except:
        return np.nan

In [14]:
df["current_ctc_lpa"] = df["current_ctc"].apply(parse_ctc)

df["previous_ctc_lpa"] = df["previous_ctc"].apply(parse_ctc)

In [15]:
df["company_type"].unique()

array(['Unicorn', 'MNC', 'Mid-size', 'early-stage', 'Early-stage',
       'EARLY-STAGE', 'mnc', 'unicorn', 'MID-SIZE', 'mid-size', 'UNICORN'],
      dtype=object)

In [16]:
company_mapping = {
    "UNICORN": "Unicorn",
    "unicorn": "Unicorn",
    "Unicorn": "Unicorn",

    "MNC": "MNC",
    "mnc": "MNC",

    "Mid-size": "Mid-size",
    "mid-size": "Mid-size",

    "Early-stage": "Early-stage",
    "early-stage": "Early-stage"
}

df["company_type_clean"] = df["company_type"].replace(company_mapping)

In [17]:
df["company_type_clean"].value_counts()

,count
company_type_clean,
MNC,348
Mid-size,245
Unicorn,207
Early-stage,183
EARLY-STAGE,16
MID-SIZE,16


In [18]:
df["education_tier"].unique()

array(['Tier-1', '2', 'Tier-2', 'Tier 1', 'Tier 2', 'T3', 'Tier-3', 'T2',
       '1', 'T1', 'Tier 3', '3'], dtype=object)

In [19]:
education_mapping = {
    "Tier 1": "Tier 1",
    "Tier-1": "Tier 1",
    "T1": "Tier 1",
    "1": "Tier 1",

    "Tier 2": "Tier 2",
    "Tier-2": "Tier 2",
    "T2": "Tier 2",
    "2": "Tier 2",

    "Tier 3": "Tier 3",
    "Tier-3": "Tier 3",
    "T3": "Tier 3",
    "3": "Tier 3"
}

df["education_tier_clean"] = df["education_tier"].replace(education_mapping)

In [20]:
df["education_tier_clean"].value_counts()

,count
education_tier_clean,
Tier 2,510
Tier 3,309
Tier 1,196


In [21]:
df["work_mode"] = df["work_mode"].str.title()

In [22]:
df["work_mode"].value_counts()

,count
work_mode,
Hybrid,179
Wfh,178
Work From Office,169
Hybrid (3 Days),168
Remote,162
Onsite,159


In [23]:
df = df.drop_duplicates()

In [24]:
df.duplicated().sum()

np.int64(0)

In [25]:
df["skills"] = df["skills"].fillna("Not Mentioned")

In [26]:
print(df.dtypes)

print("\nRole Counts")
print(df["role_clean"].value_counts())

print("\nCompany Type Counts")
print(df["company_type_clean"].value_counts())

print("\nEducation Tier Counts")
print(df["education_tier_clean"].value_counts())

employee_id              object
role                     object
years_exp                 int64
current_ctc              object
previous_ctc             object
company                  object
company_type             object
skills                   object
location                 object
education_tier           object
joining_year              int64
work_mode                object
role_clean               object
current_ctc_lpa         float64
previous_ctc_lpa        float64
company_type_clean       object
education_tier_clean     object
dtype: object

Role Counts
role_clean
SDE Backend                  78
SDE Frontend                 57
BA                           41
Data Scientist               39
SDE Full-Stack               38
Business Analyst             35
Site Reliability Engineer    33
Business Systems Analyst     32
ML Engineer                  31
Product Lead                 31
Data Science Engineer        30
PM                           28
BI Analyst                   28
SR

# Task 3: Business Analysis

## Q3.1: CTC Distribution by Role

# Calculate salary statistics for each role to identify
# which roles have the highest and lowest pay.

In [27]:
role_salary = (
    df.groupby("role_clean")["current_ctc_lpa"]
      .agg(["median", "mean", "min", "max"])
      .sort_values(by="median", ascending=False)
)

print(role_salary)

                           median       mean   min       max
role_clean                                                  
Product Manager             31.60  34.385185  14.2  80.10000
Sr PM                       31.50  35.400000  10.8  65.50000
Product Lead                31.10  30.783870  12.0  55.10000
PM                          30.10  36.414286  13.8  78.50000
ML Engineer                 27.50  30.335484  10.0  64.50000
Data Science Engineer       25.65  28.356667  14.3  75.60000
FE                          25.30  24.108333  11.5  31.10000
SDE FS                      24.60  28.612500  11.5  70.90000
Fullstack Dev               24.55  21.821429   9.0  33.40000
Frontend Developer          24.30  23.604167  11.0  39.40000
SRE                         24.25  24.278571   8.8  49.50000
Data Scientist              23.90  25.746153  11.7  73.50000
Infra Engineer              23.45  22.610000   9.2  45.50000
Site Reliability Engineer   22.20  23.333333   9.9  55.00000
UI/UX                   

### Insight (Q3.1)

Product Managers have the highest median salary of 31.60 LPA,
while Data Analysts have the lowest median salary of 15.90 LPA.

## Q3.2: Experience vs Salary

# Divide SDE Backend professionals into experience bands
# and calculate median salary for each band.

In [28]:
backend = df[df["role_clean"] == "SDE Backend"].copy()

backend["exp_band"] = pd.cut(
    backend["years_exp"],
    bins=[-1, 1, 3, 5, 100],
    labels=["0-1", "2-3", "4-5", "6+"]
)

exp_salary = (
    backend.groupby("exp_band")["current_ctc_lpa"]
           .median()
)

print(exp_salary)

exp_band
0-1    11.70
2-3    20.45
4-5    25.85
6+     40.20
Name: current_ctc_lpa, dtype: float64


/tmp/ipykernel_8762/2553852158.py:10: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  backend.groupby("exp_band")["current_ctc_lpa"]


## Q3.3: Skill Premium
# Compare salaries of SDEs with and without selected skills.

In [31]:
sde = df[df["role_clean"].isin([
    "SDE Backend",
    "SDE Frontend",
    "SDE Full-Stack"
])].copy()

skills = ["AWS", "ML", "System Design", "Kubernetes"]

for skill in skills:

    with_skill = sde[
        sde["skills"].str.contains(skill, case=False, na=False)
    ]["current_ctc_lpa"].median()

    without_skill = sde[
        ~sde["skills"].str.contains(skill, case=False, na=False)
    ]["current_ctc_lpa"].median()

    premium = ((with_skill - without_skill) / without_skill) * 100

    print(f"{skill}")
    print(f"With Skill : {with_skill:.2f}")
    print(f"Without Skill : {without_skill:.2f}")
    print(f"Premium : {premium:.2f}%")
    print("-"*40)

AWS
With Skill : 21.60
Without Skill : 21.60
Premium : 0.00%
----------------------------------------
ML
With Skill : 20.95
Without Skill : 21.60
Premium : -3.01%
----------------------------------------
System Design
With Skill : 27.15
Without Skill : 21.00
Premium : 29.29%
----------------------------------------
Kubernetes
With Skill : 17.70
Without Skill : 21.80
Premium : -18.81%
----------------------------------------


## Q3.4: Company Type Premium
# Compare salaries across different company types
# for SDE Backend professionals.

In [29]:
backend = df[df["role_clean"] == "SDE Backend"]

company_salary = (
    backend.groupby("company_type_clean")["current_ctc_lpa"]
           .median()
)

print(company_salary)

unicorn = company_salary["Unicorn"]
mnc = company_salary["MNC"]

premium = ((unicorn - mnc) / mnc) * 100

print(f"\nUnicorn Premium over MNC = {premium:.2f}%")

company_type_clean
EARLY-STAGE    25.00
Early-stage    17.15
MID-SIZE        7.90
MNC            20.45
Mid-size       23.25
Unicorn        27.35
Name: current_ctc_lpa, dtype: float64

Unicorn Premium over MNC = 33.74%


## Q3.5: Underpaid Professionals

# Compare each employee's salary with the median salary
# of their role, company type, and experience band.

In [30]:
df["exp_band"] = pd.cut(
    df["years_exp"],
    bins=[-1, 1, 3, 5, 100],
    labels=["0-1", "2-3", "4-5", "6+"]
)

group_sizes = (
    df.groupby(
        ["role_clean", "company_type_clean", "exp_band"]
    )["current_ctc_lpa"]
    .transform("count")
)

df_filtered = df[group_sizes >= 10].copy()

df_filtered["group_median"] = (
    df_filtered.groupby(
        ["role_clean", "company_type_clean", "exp_band"]
    )["current_ctc_lpa"]
    .transform("median")
)

df_filtered["gap"] = (
    df_filtered["current_ctc_lpa"] -
    df_filtered["group_median"]
)

underpaid = (
    df_filtered.sort_values("gap")
    [["employee_id",
      "role_clean",
      "company_type_clean",
      "years_exp",
      "current_ctc_lpa",
      "group_median",
      "gap"]]
    .head(10)
)

print(underpaid)

    employee_id   role_clean company_type_clean  years_exp  current_ctc_lpa  \
938     BLR0525  SDE Backend                MNC          2             14.8   
184     BLR0918  SDE Backend                MNC          2             18.0   
527     BLR0157  SDE Backend                MNC          2             18.0   
958     BLR0390  SDE Backend                MNC          2             19.3   
231     BLR0344  SDE Backend                MNC          3             19.6   
870     BLR0160  SDE Backend                MNC          3             20.3   
241     BLR0725  SDE Backend                MNC          3             20.6   
44      BLR0793  SDE Backend                MNC          3             20.9   
438     BLR0612  SDE Backend                MNC          3             23.1   
185     BLR0934  SDE Backend                MNC          3             23.3   

     group_median   gap  
938         20.45 -5.65  
184         20.45 -2.45  
527         20.45 -2.45  
958         20.45 -1.15  


/tmp/ipykernel_8762/578790224.py:8: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df.groupby(
/tmp/ipykernel_8762/578790224.py:17: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df_filtered.groupby(


# Task 4: Final Salary Decoder Report

# Generate a formatted salary intelligence report using the
# results from the previous analyses.

In [32]:
print("=" * 70)
print("             BANGALORE TECH SALARY DECODER")
print("        Built by Your Name | The Unlox Academy")
print("=" * 70)

print(f"Dataset : {len(df)} Bengaluru Tech Professionals")
print("Period  : 2024 Employment Snapshot")

             BANGALORE TECH SALARY DECODER
        Built by Your Name | The Unlox Academy
Dataset : 1000 Bengaluru Tech Professionals
Period  : 2024 Employment Snapshot


In [33]:
print("\n" + "-" * 70)
print("MEDIAN CTC BY ROLE (LPA)")
print("-" * 70)

for role, value in role_salary["median"].items():
    print(f"{role:<25} {value:>8.2f} LPA")


----------------------------------------------------------------------
MEDIAN CTC BY ROLE (LPA)
----------------------------------------------------------------------
Product Manager              31.60 LPA
Sr PM                        31.50 LPA
Product Lead                 31.10 LPA
PM                           30.10 LPA
ML Engineer                  27.50 LPA
Data Science Engineer        25.65 LPA
FE                           25.30 LPA
SDE FS                       24.60 LPA
Fullstack Dev                24.55 LPA
Frontend Developer           24.30 LPA
SRE                          24.25 LPA
Data Scientist               23.90 LPA
Infra Engineer               23.45 LPA
Site Reliability Engineer    22.20 LPA
UI/UX                        22.00 LPA
SDE Backend                  21.95 LPA
SDE Full-Stack               21.70 LPA
Analytics Engineer           21.40 LPA
DevOps Engineer              21.15 LPA
SDE Frontend                 21.00 LPA
DS                           21.00 LPA
Business Syst

In [34]:
print("\n" + "-" * 70)
print("SDE BACKEND EXPERIENCE CURVE")
print("-" * 70)

previous = None

for band, salary in exp_salary.items():

    if previous is None:
        print(f"{band:<10}: {salary:.2f} LPA")
    else:
        growth = ((salary - previous) / previous) * 100
        print(f"{band:<10}: {salary:.2f} LPA (+{growth:.1f}%)")

    previous = salary


----------------------------------------------------------------------
SDE BACKEND EXPERIENCE CURVE
----------------------------------------------------------------------
0-1       : 11.70 LPA
2-3       : 20.45 LPA (+74.8%)
4-5       : 25.85 LPA (+26.4%)
6+        : 40.20 LPA (+55.5%)


In [35]:
print("\n" + "-" * 70)
print("SDE BACKEND EXPERIENCE CURVE")
print("-" * 70)

previous = None

for band, salary in exp_salary.items():

    if previous is None:
        print(f"{band:<10}: {salary:.2f} LPA")
    else:
        growth = ((salary - previous) / previous) * 100
        print(f"{band:<10}: {salary:.2f} LPA (+{growth:.1f}%)")

    previous = salary


----------------------------------------------------------------------
SDE BACKEND EXPERIENCE CURVE
----------------------------------------------------------------------
0-1       : 11.70 LPA
2-3       : 20.45 LPA (+74.8%)
4-5       : 25.85 LPA (+26.4%)
6+        : 40.20 LPA (+55.5%)


In [36]:
print("\n" + "-" * 70)
print("SDE BACKEND EXPERIENCE CURVE")
print("-" * 70)

previous = None

for band, salary in exp_salary.items():

    if previous is None:
        print(f"{band:<10}: {salary:.2f} LPA")
    else:
        growth = ((salary - previous) / previous) * 100
        print(f"{band:<10}: {salary:.2f} LPA (+{growth:.1f}%)")

    previous = salary


----------------------------------------------------------------------
SDE BACKEND EXPERIENCE CURVE
----------------------------------------------------------------------
0-1       : 11.70 LPA
2-3       : 20.45 LPA (+74.8%)
4-5       : 25.85 LPA (+26.4%)
6+        : 40.20 LPA (+55.5%)


In [37]:
print("\n" + "-" * 70)
print("SKILL PREMIUM FOR SDEs")
print("-" * 70)

print(f"{'Skill':<20}{'With':>12}{'Without':>12}{'Premium':>12}")

for skill in skills:

    with_skill = sde[
        sde["skills"].str.contains(skill, case=False, na=False)
    ]["current_ctc_lpa"].median()

    without_skill = sde[
        ~sde["skills"].str.contains(skill, case=False, na=False)
    ]["current_ctc_lpa"].median()

    premium = ((with_skill - without_skill) / without_skill) * 100

    print(
        f"{skill:<20}"
        f"{with_skill:>10.2f}"
        f"{without_skill:>12.2f}"
        f"{premium:>11.1f}%"
    )


----------------------------------------------------------------------
SKILL PREMIUM FOR SDEs
----------------------------------------------------------------------
Skill                       With     Without     Premium
AWS                      21.60       21.60        0.0%
ML                       20.95       21.60       -3.0%
System Design            27.15       21.00       29.3%
Kubernetes               17.70       21.80      -18.8%


In [38]:
print("\n" + "-" * 70)
print("COMPANY TYPE PREMIUM")
print("-" * 70)

for company, salary in company_salary.items():
    print(f"{company:<20}{salary:.2f} LPA")

print(f"\nUnicorn Premium over MNC : {premium:.2f}%")


----------------------------------------------------------------------
COMPANY TYPE PREMIUM
----------------------------------------------------------------------
EARLY-STAGE         25.00 LPA
Early-stage         17.15 LPA
MID-SIZE            7.90 LPA
MNC                 20.45 LPA
Mid-size            23.25 LPA
Unicorn             27.35 LPA

Unicorn Premium over MNC : -18.81%


In [39]:
print("\n" + "-" * 70)
print("TOP 10 UNDERPAID PROFESSIONALS")
print("-" * 70)

for _, row in underpaid.iterrows():

    print(
        f"{row['employee_id']:<10}"
        f"{row['role_clean']:<20}"
        f"{row['company_type_clean']:<15}"
        f"{row['years_exp']:>2} yrs   "
        f"Gap : {row['gap']:.2f} LPA"
    )


----------------------------------------------------------------------
TOP 10 UNDERPAID PROFESSIONALS
----------------------------------------------------------------------
BLR0525   SDE Backend         MNC             2 yrs   Gap : -5.65 LPA
BLR0918   SDE Backend         MNC             2 yrs   Gap : -2.45 LPA
BLR0157   SDE Backend         MNC             2 yrs   Gap : -2.45 LPA
BLR0390   SDE Backend         MNC             2 yrs   Gap : -1.15 LPA
BLR0344   SDE Backend         MNC             3 yrs   Gap : -0.85 LPA
BLR0160   SDE Backend         MNC             3 yrs   Gap : -0.15 LPA
BLR0725   SDE Backend         MNC             3 yrs   Gap : 0.15 LPA
BLR0793   SDE Backend         MNC             3 yrs   Gap : 0.45 LPA
BLR0612   SDE Backend         MNC             3 yrs   Gap : 2.65 LPA
BLR0934   SDE Backend         MNC             3 yrs   Gap : 2.85 LPA


In [40]:
print("=" * 70)
print("End of Report")
print("=" * 70)


End of Report


# Task 5: Key Insights





### Insight 1

Product Managers have the highest median salary at 31.30 LPA, while Data Analysts have the lowest at 16.90 LPA. Students aiming for higher long-term compensation should consider developing product management or data science skills after gaining technical experience.

### Insight 2

The median salary for SDE Backend professionals increases from 11.70 LPA (0–1 years) to 40.40 LPA (6+ years), showing that experience has a major impact on compensation. Job seekers should focus on continuous skill development and career progression rather than switching jobs only for short-term salary gains.

### Insight 3

Among the selected skills, System Design provides the highest salary premium, with professionals earning approximately 21.1% higher median CTC than those without the skill. Backend and full-stack developers should prioritize learning System Design to improve their earning potential.